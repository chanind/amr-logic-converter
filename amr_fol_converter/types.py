from __future__ import annotations
from dataclasses import dataclass
from typing import Any, Union


@dataclass
class Const:
    node: Any

    def __str__(self) -> str:
        return str(self.node)


@dataclass
class Var:
    name: str

    def __str__(self) -> str:
        return self.name


@dataclass
class Predicate:
    node: Any
    args: tuple[Const | Var, ...]

    def __str__(self) -> str:
        return f"{str(self.node)}({', '.join(map(str, self.args))})"


@dataclass
class And:
    args: tuple["Formula", ...]

    def __str__(self) -> str:
        return f"{' ^ '.join(map(str, self.args))}"


@dataclass
class Not:
    body: "Formula"

    def __str__(self) -> str:
        return f"¬{str(self.body)}"


@dataclass
class Exists:
    var: Var
    body: "Formula"

    def __str__(self) -> str:
        return f"∃{str(self.var)}({str(self.body)})"


Formula = Union[Const, Var, Predicate, Not, Exists, And]
