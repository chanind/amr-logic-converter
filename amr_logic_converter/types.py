from __future__ import annotations
from dataclasses import dataclass
from typing import Union
from typing_extensions import Literal
from penman.surface import Alignment

from .parse_value_and_alignment import parse_value_and_alignment


ConstType = Literal["string", "symbol", "instance"]


@dataclass
class Const:
    value: str
    type: ConstType
    alignment: Alignment | None = None

    def __init__(self, element: str, type: ConstType) -> None:
        self.type = type
        self.value, self.alignment = parse_value_and_alignment(element)
        # remove explicit quotes from string literals
        if (
            self.type == "string"
            and self.value.startswith('"')
            and self.value.endswith('"')
        ):
            self.value = self.value[1:-1]

    def __str__(self) -> str:
        if self.type == "string":
            return f'"{self.value}"'
        return self.value


@dataclass
class Param:
    name: str

    def __str__(self) -> str:
        return self.name


@dataclass
class Predicate:
    value: str
    args: tuple[Const | Param, ...]
    alignment: Alignment | None = None

    def __init__(self, element: str, args: tuple[Const | Param, ...]) -> None:
        self.value, self.alignment = parse_value_and_alignment(element)
        self.args = args

    def __str__(self) -> str:
        return f"{str(self.value)}({', '.join(map(str, self.args))})"


@dataclass
class And:
    args: tuple["Formula", ...]

    def __init__(self, args: tuple["Formula", ...]) -> None:
        # automatically reduce repeated ANDs
        simplified_args: list["Formula"] = []
        for arg in args:
            if type(arg) is And:
                simplified_args.extend(arg.args)
            else:
                simplified_args.append(arg)
        self.args = tuple(simplified_args)

    def __str__(self) -> str:
        arg_strs = []
        for arg in self.args:
            if type(arg) in [Or, Implies]:
                arg_strs.append(f"({str(arg)})")
            else:
                arg_strs.append(str(arg))
        return f"{' ∧ '.join(arg_strs)}"


@dataclass
class Or:
    args: tuple["Formula", ...]

    def __init__(self, args: tuple["Formula", ...]) -> None:
        # automatically reduce repeated ORs
        simplified_args: list["Formula"] = []
        for arg in args:
            if type(arg) is Or:
                simplified_args.extend(arg.args)
            else:
                simplified_args.append(arg)
        self.args = tuple(simplified_args)

    def __str__(self) -> str:
        arg_strs = []
        for arg in self.args:
            if type(arg) in [And, Implies]:
                arg_strs.append(f"({str(arg)})")
            else:
                arg_strs.append(str(arg))
        return f"{' ∨ '.join(arg_strs)}"


@dataclass
class Not:
    body: "Formula"

    def __str__(self) -> str:
        if type(self.body) is And:
            return f"¬({str(self.body)})"
        return f"¬{str(self.body)}"


@dataclass
class Implies:
    antecedent: "Formula"
    consequent: "Formula"

    def __str__(self) -> str:
        antecedent_str = str(self.antecedent)
        consequent_str = str(self.consequent)
        if type(self.antecedent) in [And, Or, Implies]:
            antecedent_str = f"({antecedent_str})"
        if type(self.consequent) in [And, Or, Implies]:
            consequent_str = f"({consequent_str})"
        return f"{antecedent_str} → {consequent_str}"


@dataclass
class Exists:
    param: Param
    body: "Formula"

    def __str__(self) -> str:
        return f"∃{str(self.param)}({str(self.body)})"


Formula = Union[Predicate, Not, Exists, And, Or]
