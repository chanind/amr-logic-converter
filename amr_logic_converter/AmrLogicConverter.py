from __future__ import annotations
from dataclasses import dataclass

from functools import reduce
from typing import Callable, Optional, Union, cast

import penman
from penman.tree import Tree, Node, Branch
from penman.graph import Graph

from amr_logic_converter.AmrContext import (
    AmrContext,
    OverrideIsProjectiveCallback,
    OverrideIsProjectiveCallbackInfo,
)
from amr_logic_converter.types import (
    And,
    Constant,
    ConstantType,
    Exists,
    Clause,
    Implies,
    Not,
    Predicate,
    Variable,
    Atom,
)


@dataclass
class OverrideQuantificationCallbackInfo:
    """Metadata passed to the OverrideQuantificationCallback wtih info about the node being processed"""

    instance_name: str
    bound_instance: Variable | Constant
    node: Node
    depth: int
    amr_tree: Tree
    is_negated: bool


OverrideQuantificationCallback = Callable[
    [Clause, OverrideQuantificationCallbackInfo], Union[Clause, None]
]


@dataclass
class OverrideConjunctionCallbackInfo:
    """Metadata passed to the OverrideConjunctionCallback wtih info about the node being processed"""

    predicate_term: Atom
    closure_term: Clause | None
    subterms: list[Clause]
    condition_term: Clause | None
    instance_name: str
    bound_instance: Variable | Constant
    node: Node
    depth: int
    amr_tree: Tree


OverrideConjunctionCallback = Callable[
    [Clause, OverrideConjunctionCallbackInfo], Union[Clause, None]
]


def normalize_atom(atom: Atom) -> Atom:
    # flip :ARGX-of(x,y) to :ARGX(y,x)
    if atom.symbol.endswith("-of") and len(atom.terms) == 2:
        predicate = Predicate(atom.symbol[:-3], atom.predicate.alignment)
        return predicate(atom.terms[1], atom.terms[0])
    return atom


def determine_const_type(value: str) -> ConstantType:
    return "string" if value.startswith('"') else "symbol"


class AmrLogicConverter:
    """
    Main entry point for converting AMR to a logic formula.

    basic usage:
    converter = AmrLogicConverter()
    logic = converter.convert(
        '(c / chase-01~4'
        '   :ARG0~5 (d / dog~7)'
        '   :ARG0~3 (c / cat~2))'
    )
    print(logic)
    """

    invert_relations: bool
    existentially_quantify_instances: bool
    use_variables_for_instances: bool
    maximally_hoist_coreferences: bool
    capitalize_variables: bool
    use_implies_for_conditions: bool
    override_is_projective: Optional[OverrideIsProjectiveCallback]
    override_quantification: Optional[OverrideQuantificationCallback]
    override_conjunction: Optional[OverrideConjunctionCallback]

    def __init__(
        self,
        invert_relations: bool = True,
        existentially_quantify_instances: bool = False,
        use_variables_for_instances: bool = False,
        maximally_hoist_coreferences: bool = False,
        capitalize_variables: bool = True,
        use_implies_for_conditions: bool = False,
        override_is_projective: Optional[OverrideIsProjectiveCallback] = None,
        override_quantification: Optional[OverrideQuantificationCallback] = None,
        override_conjunction: Optional[OverrideConjunctionCallback] = None,
    ) -> None:
        self.invert_relations = invert_relations
        self.capitalize_variables = capitalize_variables
        self.existentially_quantify_instances = existentially_quantify_instances
        self.use_variables_for_instances = use_variables_for_instances
        self.use_implies_for_conditions = use_implies_for_conditions
        self.maximally_hoist_coreferences = maximally_hoist_coreferences
        self.override_is_projective = override_is_projective
        self.override_quantification = override_quantification
        self.override_conjunction = override_conjunction

    def _get_bound_instance(self, instance_name: str) -> Variable | Constant:
        use_variables_for_instances = (
            self.existentially_quantify_instances or self.use_variables_for_instances
        )
        bound_instance: Variable | Constant = (
            Variable(self._var_name(instance_name))
            if use_variables_for_instances
            else Constant(instance_name, "instance")
        )
        return bound_instance

    def _reprioritize_edge(self, edge: Branch) -> int:
        """Reprioritize edges to make it possible to change where in the tree coreferenced instances are defined"""
        role = edge[0]
        # make sure :condition is processed first if we're rewriting using Implies
        if self.use_implies_for_conditions and role == ":condition":
            return 1
        return 0

    def _sort_edges(self, edges: list[Branch], reverse: bool = True) -> list[Branch]:
        """Sort edges by reprioritized edge priority"""
        return sorted(edges, key=self._reprioritize_edge, reverse=reverse)

    def _var_name(self, name: str) -> str:
        return name.capitalize() if self.capitalize_variables else name

    def _convert_amr_assertive(
        self,
        ctx: AmrContext,
        instance_name: str,
        closure: Optional[Callable[[str], Clause | None]] = None,
    ) -> Clause | None:
        # handle 7.2, 7.6-7.8 from "Expressive Power of Abstract Meaning Representations"
        # ∥x,φ∥↓ = φ(x)
        # ∥(x\P),φ∥↓ = φ(x)
        # ∥(x\P :RiAi),φ∥↓ = φ(x)
        # ∥(x\P :RiAi :polarity–),φ∥↓ = φ(x)
        if ctx.is_instance_rendered(instance_name):
            return None if closure is None else closure(instance_name)
        node = ctx.get_node_for_instance(instance_name)
        instance_predicate, *edges = node[1]
        bound_instance = self._get_bound_instance(instance_name)
        predicate = Predicate.from_amr_str(instance_predicate[1])
        predicate_term = predicate(bound_instance)
        closure_term = closure(instance_name) if closure is not None else None
        subterms: list[Clause] = []
        condition_term: Clause | None = None

        ctx.mark_instance_rendered(instance_name)
        for (role, target) in self._sort_edges(edges):

            def sub_closure(u: str) -> Atom:
                target: Variable | Constant = self._get_bound_instance(u)
                sub_predicate = Predicate.from_amr_str(role)
                atom = sub_predicate(bound_instance, target)
                return normalize_atom(atom) if self.invert_relations else atom

            # don't include the :condition relation in the logic if we're turning it into an implication
            target_closure: Callable[[str], Atom] | None = sub_closure
            if role == ":condition" and self.use_implies_for_conditions:
                target_closure = None

            target_instance = _get_instance_name(target, ctx)
            subterm: Clause | None = None
            # special case for the :polarity - attribute.
            # skip polarity as is handled in quantification
            if role == ":polarity" and target == "-":
                continue
            elif target_instance is not None:
                target_node = ctx.get_node_for_instance(target_instance)
                subterm = self._convert_amr(ctx, target_node, target_closure)
            else:
                sub_predicate = Predicate.from_amr_str(role)
                sub_atom = sub_predicate(
                    bound_instance, Constant(target, determine_const_type(target))
                )
                subterm = (
                    normalize_atom(sub_atom) if self.invert_relations else sub_atom
                )
            if role == ":condition":
                condition_term = subterm
            else:
                subterms.append(subterm)

        if self.override_conjunction is not None:
            info = OverrideConjunctionCallbackInfo(
                predicate_term=predicate_term,
                closure_term=closure_term,
                subterms=subterms,
                condition_term=condition_term,
                instance_name=instance_name,
                bound_instance=bound_instance,
                depth=ctx.get_instance_depth(instance_name),
                node=node,
                amr_tree=ctx.amr_tree,
            )
            override_result = self.override_conjunction(predicate_term, info)
            if override_result is not None:
                return override_result

        pre_terms = []
        if closure_term is not None:
            pre_terms.append(closure_term)
        if condition_term is not None and not self.use_implies_for_conditions:
            pre_terms.append(condition_term)
        pre_terms.append(predicate_term)
        conjunction = And(*(pre_terms + subterms))
        if condition_term is not None and self.use_implies_for_conditions:
            return Implies(condition_term, conjunction)
        return conjunction

    def _quantify_instance(
        self,
        ctx: AmrContext,
        instance_name: str,
    ) -> Callable[[Clause], Clause]:
        node = ctx.get_node_for_instance(instance_name)
        _instance_predicate, *edges = node[1]
        bound_instance = self._get_bound_instance(instance_name)
        polarity = True
        for role, target in self._sort_edges(edges):
            if role == ":polarity" and target == "-":
                polarity = False

        def quantification_closure(clause: Clause) -> Clause:
            if self.override_quantification is not None:
                override_expr = self.override_quantification(
                    clause,
                    OverrideQuantificationCallbackInfo(
                        node=node,
                        instance_name=instance_name,
                        bound_instance=bound_instance,
                        depth=ctx.get_instance_depth(instance_name),
                        amr_tree=ctx.amr_tree,
                        is_negated=not polarity,
                    ),
                )
                if override_expr is not None:
                    return override_expr
            expr = clause
            if self.existentially_quantify_instances:
                expr = Exists(cast(Variable, bound_instance), body=clause)
            return expr if polarity else Not(expr)

        return quantification_closure

    def _quanitfy_formula(
        self, ctx: AmrContext, formula: Clause, instances: set[str]
    ) -> Clause:
        """Wrap the formula in quantifiers for all instances in the list"""
        sorted_instances = sorted(
            sorted(instances),  # sort alphabetically as a tie-breaker
            key=lambda instance: ctx.get_instance_depth(instance),
            reverse=True,
        )

        return reduce(
            lambda formula, instance_name: self._quantify_instance(ctx, instance_name)(
                formula
            ),
            sorted_instances,
            formula,
        )

    def _convert_amr_projective(
        self,
        ctx: AmrContext,
        node: Node,
        context_node: Node | None,
    ) -> Callable[[Clause | None], Clause | None]:
        instance_name, instance_info = node
        # only project instances at their specific scope
        projections_for_context = ctx.get_instances_at_scope(context_node)

        edges = instance_info[1:]
        cur_closure = lambda x: x
        if instance_name in projections_for_context:
            # handle 8.6-8.8 from "Expressive Power of Abstract Meaning Representations"
            # ∥(x\P :RiAi)∥↑ = λp.∥(x/P :RiAi),λx.p∥↓
            cur_closure = lambda u: self._convert_amr_assertive(
                ctx, instance_name, lambda x: u
            )

        def args_closure(p: Clause | None) -> Clause | None:
            # handle 8.3-8.5 from "Expressive Power of Abstract Meaning Representations"
            # ∥(x/P :RiAi)∥↑ = λp.∥A1∥↑(∥A2∥↑( ...∥An∥↑(p)))
            # don't need to worry about iterating over non-nodes since those are just λp.p
            result = cur_closure(p)
            for edge in self._sort_edges(edges, reverse=False):
                if type(edge[1]) is tuple:
                    sub_closure = self._convert_amr_projective(
                        ctx, edge[1], context_node
                    )
                    result = sub_closure(result)
            return result

        return args_closure

    def _convert_amr(
        self,
        ctx: AmrContext,
        node: Node,
        assertive_closure: Optional[Callable[[str], Clause]] = None,
    ) -> Clause:
        instances_to_quantify = ctx.get_instances_to_quantify_at_scope(node)
        ctx.mark_instances_quantified(instances_to_quantify)
        projective_closure = self._convert_amr_projective(ctx, node, node)
        base_formula = cast(
            Clause,
            projective_closure(
                self._convert_amr_assertive(ctx, node[0], assertive_closure)
            ),
        )
        return self._quanitfy_formula(ctx, base_formula, instances_to_quantify)

    def _maximally_project_amr(
        self, ctx: AmrContext, node: Node
    ) -> Callable[[Clause], Clause]:
        return cast(
            Callable[[Clause], Clause], self._convert_amr_projective(ctx, node, None)
        )

    def _override_is_projective(
        self, info: OverrideIsProjectiveCallbackInfo
    ) -> bool | None:
        override: bool | None = None
        if self.maximally_hoist_coreferences and info.is_coreferent:
            override = True
        if self.override_is_projective is not None:
            callback_res = self.override_is_projective(info)
            if callback_res is not None:
                override = callback_res
        return override

    def convert_amr_tree(self, amr_tree: Tree) -> Clause:
        ctx = AmrContext.from_amr_tree(
            amr_tree, override_is_projective=self._override_is_projective
        )

        # special case to handle maximally projected instances
        maximal_projection = self._maximally_project_amr(ctx, amr_tree.node)
        formula = maximal_projection(self._convert_amr(ctx, amr_tree.node))
        maximum_scope_instances = ctx.get_instances_at_scope(None)
        return self._quanitfy_formula(ctx, formula, maximum_scope_instances)

    def convert_amr_str(self, amr_str: str) -> Clause:
        return self.convert_amr_tree(penman.parse(amr_str))

    def convert(self, amr: str | Tree | Graph) -> Clause:
        if isinstance(amr, str):
            return self.convert_amr_str(amr)
        elif isinstance(amr, Tree):
            return self.convert_amr_tree(amr)
        elif isinstance(amr, Graph):
            return self.convert_amr_tree(penman.configure(amr))
        else:
            raise TypeError(
                f"Expected amr to be a string, Tree, or Graph. Got {type(amr)}"
            )


def _get_instance_name(target: Node | str, ctx: AmrContext) -> str | None:
    if type(target) is tuple:
        return target[0]
    if target in ctx.instances:
        return target
    return None
