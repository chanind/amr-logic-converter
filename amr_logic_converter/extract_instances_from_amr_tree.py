from __future__ import annotations

import penman
from penman.models import amr


def extract_instances_from_amr_tree(amr_tree: penman.Tree) -> frozenset[str]:
    """Extract the set of instances in the given AMR tree."""
    instances = set()
    amr_graph = penman.interpret(amr_tree, model=amr.model)
    for edge in amr_graph.edges():
        instances.add(edge.source)
        instances.add(edge.target)
    return frozenset(instances)
