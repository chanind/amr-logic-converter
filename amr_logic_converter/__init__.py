__version__ = "0.3.1"

from .AmrLogicConverter import AmrLogicConverter
from .types import And, Or, Const, Exists, Formula, Not, Predicate, Param, Implies

__all__ = [
    "AmrLogicConverter",
    "And",
    "Or",
    "Const",
    "Exists",
    "Formula",
    "Implies",
    "Not",
    "Predicate",
    "Param",
]
