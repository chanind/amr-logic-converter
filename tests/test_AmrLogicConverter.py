from __future__ import annotations
from typing import cast

import penman
import pytest
from syrupy.assertion import SnapshotAssertion

from amr_logic_converter import AmrLogicConverter
from amr_logic_converter.AmrLogicConverter import (
    OverrideConjunctionCallbackInfo,
    OverrideQuantificationCallbackInfo,
)
from amr_logic_converter.types import All, And, Clause, Implies, Not, Variable
from tests.test_utils import fmt_logic


converter = AmrLogicConverter(
    existentially_quantify_instances=True, capitalize_variables=False
)


def test_convert_basic_amr() -> None:
    amr_str = """
    (e / give-01
        :ARG0 (x / person :named "Ms Ribble")
        :ARG2 (y / child)
        :ARG1 (z / envelope))
    """
    expected = '∃e(give-01(e) ∧ ∃x(:ARG0(e, x) ∧ person(x) ∧ :named(x, "Ms Ribble")) ∧ ∃y(:ARG2(e, y) ∧ child(y)) ∧ ∃z(:ARG1(e, z) ∧ envelope(z)))'
    logic = converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_works_with_amr_tree() -> None:
    amr_str = """
    (e / give-01
        :ARG0 (x / person :named "Ms Ribble")
        :ARG1 (z / envelope))
    """
    expected = '∃e(give-01(e) ∧ ∃x(:ARG0(e, x) ∧ person(x) ∧ :named(x, "Ms Ribble")) ∧ ∃z(:ARG1(e, z) ∧ envelope(z)))'
    logic = converter.convert(penman.parse(amr_str))
    assert str(logic) == expected


def test_convert_works_with_amr_graph() -> None:
    amr_str = """
    (e / give-01
        :ARG0 (x / person :named "Ms Ribble")
        :ARG1 (z / envelope))
    """
    expected = '∃e(give-01(e) ∧ ∃x(:ARG0(e, x) ∧ person(x) ∧ :named(x, "Ms Ribble")) ∧ ∃z(:ARG1(e, z) ∧ envelope(z)))'
    logic = converter.convert(penman.decode(amr_str))
    assert str(logic) == expected


def test_convert_errors_on_invalid_input() -> None:
    with pytest.raises(TypeError):
        converter.convert(17)


def test_convert_basic_amr_with_role_inversion() -> None:
    amr_str = """
    (y / book
        :ARG1-of (e / read-01
            :ARG0 (x / girl)))
    """
    expected = "∃y(book(y) ∧ ∃e(:ARG1(e, y) ∧ read-01(e) ∧ ∃x(:ARG0(e, x) ∧ girl(x))))"
    logic = converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_capitalized_variables_by_default() -> None:
    amr_str = """
    (y / book
        :ARG1-of (e / read-01
            :ARG0 (x / girl)))
    """
    capitalized_converter = AmrLogicConverter(existentially_quantify_instances=True)
    expected = "∃Y(book(Y) ∧ ∃E(:ARG1(E, Y) ∧ read-01(E) ∧ ∃X(:ARG0(E, X) ∧ girl(X))))"
    logic = capitalized_converter.convert(amr_str)
    assert str(logic) == expected


def test__convert_amr_skips_role_inversion_if_specified() -> None:
    amr_str = """
    (y / book
        :ARG1-of (e / read-01
            :ARG0 (x / girl)))
    """
    expected = (
        "∃y(book(y) ∧ ∃e(:ARG1-of(y, e) ∧ read-01(e) ∧ ∃x(:ARG0(e, x) ∧ girl(x))))"
    )
    no_inversion_converter = AmrLogicConverter(
        invert_relations=False,
        existentially_quantify_instances=True,
        capitalize_variables=False,
    )
    logic = no_inversion_converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_with_negation() -> None:
    amr_str = """
    (e / giggle-01
        :polarity -
        :ARG0 (x / boy))
    """
    expected = "¬∃e(giggle-01(e) ∧ ∃x(:ARG0(e, x) ∧ boy(x)))"
    logic = converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_with_negation_maintains_negation_scope_when_inverted() -> None:
    amr_str = """
    (x / boy
        :ARG0-of (e / giggle-01
            :polarity -))
    """
    expected = "∃x(boy(x) ∧ ¬∃e(:ARG0(e, x) ∧ giggle-01(e)))"
    logic = converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_with_coreference_and_no_subattrs() -> None:
    amr_str = """
    (e / dry-01
        :ARG0 (x / person)
        :ARG1 x)
    """

    expected = "∃e(∃x(dry-01(e) ∧ :ARG0(e, x) ∧ person(x) ∧ :ARG1(e, x)))"
    logic = converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_with_coreference() -> None:
    amr_str = """
    (e / dry-01
        :ARG0 (x / person
            :named "Mr Krupp")
        :ARG1 x)
    """

    expected = '∃e(∃x(dry-01(e) ∧ :ARG0(e, x) ∧ person(x) ∧ :named(x, "Mr Krupp") ∧ :ARG1(e, x)))'
    logic = converter.convert(amr_str)
    print(logic)
    assert str(logic) == expected


def test_convert_amr_with_alignment_markers(snapshot: SnapshotAssertion) -> None:
    amr_str = """
    (e / give-01~2
        :ARG0 (x / person :named "Ms Ribble"~2)
        :ARG2 (y / child~3)
        :ARG1 (z / envelope~4))
    """
    expected = '∃e(give-01(e) ∧ ∃x(:ARG0(e, x) ∧ person(x) ∧ :named(x, "Ms Ribble")) ∧ ∃y(:ARG2(e, y) ∧ child(y)) ∧ ∃z(:ARG1(e, z) ∧ envelope(z)))'
    logic = converter.convert(amr_str)
    assert str(logic) == expected
    # alignments are included in the logic elements, just using a snapshot assert for convenience
    assert logic == snapshot


def test_convert_amr_with_nested_coreference() -> None:
    amr_str = """
    (e / dry-01
        :ARG0 (x / person
            :ARG0-of (g / giggle-01
                :polarity - ))
        :ARG1 (
            z / dog
                :ARG0-of (w / wash-01
                    :ARG1 z)))
    """
    expected = "∃e(dry-01(e) ∧ ∃x(:ARG0(e, x) ∧ person(x) ∧ ¬∃g(:ARG0(g, x) ∧ giggle-01(g))) ∧ ∃z(:ARG1(e, z) ∧ dog(z) ∧ ∃w(:ARG0(w, z) ∧ wash-01(w) ∧ :ARG1(w, z))))"
    logic = converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_with_only_coreference() -> None:
    amr_str = """
    (s / smurf
        :ARG0 s
        :ARG1 s)
    """
    expected = "∃s(smurf(s) ∧ :ARG0(s, s) ∧ :ARG1(s, s))"
    logic = converter.convert(amr_str)
    print(logic)
    assert str(logic) == expected


def test_convert_amr_leaves_off_existential_quantifiers_by_default(
    snapshot: SnapshotAssertion,
) -> None:
    amr_str = """
    (e / dry-01
        :ARG0 (x / person
            :ARG0-of (g / giggle-01
                :polarity - ))
        :ARG1 (
            z / dog
                :ARG0-of (w / wash-01
                    :ARG1 z)))
    """
    non_quantified_converter = AmrLogicConverter()
    expected = "dry-01(e) ∧ :ARG0(e, x) ∧ person(x) ∧ ¬(:ARG0(g, x) ∧ giggle-01(g)) ∧ :ARG1(e, z) ∧ dog(z) ∧ :ARG0(w, z) ∧ wash-01(w) ∧ :ARG1(w, z)"
    logic = non_quantified_converter.convert(amr_str)
    assert str(logic) == expected
    assert logic == snapshot


def test_convert_amr_can_replace_instances_with_variables(
    snapshot: SnapshotAssertion,
) -> None:
    amr_str = """
    (e / dry-01
        :ARG0 (x / person
            :ARG0-of (g / giggle-01
                :polarity - ))
        :ARG1 (
            z / dog
                :ARG0-of (w / wash-01
                    :ARG1 z)))
    """
    variables_converter = AmrLogicConverter(
        use_variables_for_instances=True,
        capitalize_variables=False,
    )
    expected = "dry-01(e) ∧ :ARG0(e, x) ∧ person(x) ∧ ¬(:ARG0(g, x) ∧ giggle-01(g)) ∧ :ARG1(e, z) ∧ dog(z) ∧ :ARG0(w, z) ∧ wash-01(w) ∧ :ARG1(w, z)"
    logic = variables_converter.convert(amr_str)
    assert str(logic) == expected
    assert logic == snapshot


def test_convert_amr_with_top_level_negation_and_deep_nesting() -> None:
    amr_str = """
    (b / bad-07~2
        :polarity -
        :ARG1 (e / dry-01
            :ARG0 (x / person
                :named "Mr Krupp")
            :ARG1 x))
    """
    expected = '¬∃b(bad-07(b) ∧ ∃e(∃x(:ARG1(b, e) ∧ dry-01(e) ∧ :ARG0(e, x) ∧ person(x) ∧ :named(x, "Mr Krupp") ∧ :ARG1(e, x))))'
    logic = converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_moves_coreferent_vars_to_widest_scope_with_maximally_hoist_coreferences_option() -> None:
    amr_str = """
    (b / bad-07~2
        :polarity -
        :ARG1 (e / dry-01
            :ARG0 (x / person
                :named "Mr Krupp")
            :ARG1 x))
    """
    expected = '∃x(¬∃b(bad-07(b) ∧ ∃e(:ARG1(b, e) ∧ dry-01(e) ∧ :ARG0(e, x) ∧ person(x) ∧ :named(x, "Mr Krupp") ∧ :ARG1(e, x))))'
    max_hoist_converter = AmrLogicConverter(
        existentially_quantify_instances=True,
        maximally_hoist_coreferences=True,
        capitalize_variables=False,
    )
    logic = max_hoist_converter.convert(amr_str)
    assert str(logic) == expected


def test_convert_amr_allows_overriding_scope_of_instances() -> None:
    amr_str = """
    (b / bad-07~2
        :polarity -
        :ARG1 (e / dry-01
            :ARG0 (x / person
                :named "Mr Krupp")
            :ARG1 x))
    """
    expected = '∃E(¬∃B(bad-07(B) ∧ ∃X(:ARG1(B, E) ∧ dry-01(E) ∧ :ARG0(E, X) ∧ person(X) ∧ :named(X, "Mr Krupp") ∧ :ARG1(E, X))))'
    override_scope_converter = AmrLogicConverter(
        existentially_quantify_instances=True,
        override_is_projective=lambda info: True if info.instance_name == "e" else None,
    )
    logic = override_scope_converter.convert(amr_str)
    print(logic)
    assert str(logic) == expected


def test_convert_amr_allows_overriding_quantification() -> None:
    amr_str = """
    (b / bad-07~2
        :polarity -
        :ARG1 (e / dry-01
            :ARG0 (x / person
                :named "Mr Krupp")
            :ARG1 x))
    """
    expected = '¬∃B(bad-07(B) ∧ ∀E(∃X(:ARG1(B, E) ∧ dry-01(E) ∧ :ARG0(E, X) ∧ person(X) ∧ :named(X, "Mr Krupp") ∧ :ARG1(E, X))))'
    override_scope_converter = AmrLogicConverter(
        existentially_quantify_instances=True,
        override_quantification=lambda clause, info: All(
            cast(Variable, info.bound_instance), clause
        )
        if info.instance_name == "e"
        else None,
    )
    logic = override_scope_converter.convert(amr_str)
    print(logic)
    assert str(logic) == expected


def test_convert_amr_allows_overriding_conjunction() -> None:
    amr_str = """
    (b / bad-07~2
        :polarity -
        :ARG1 (e / dry-01
            :ARG0 (x / person
                :named "Mr Krupp")
            :ARG1 x))
    """
    expected_with_quantifiers = '¬((∃E(∃X(:ARG1(B, E) ∧ dry-01(E) ∧ :ARG0(E, X) ∧ person(X) ∧ :named(X, "Mr Krupp") ∧ :ARG1(E, X)))) → bad-07(B))'
    expected_without_quantifiers = '¬((:ARG1(B, E) ∧ dry-01(E) ∧ :ARG0(E, X) ∧ person(X) ∧ :named(X, "Mr Krupp") ∧ :ARG1(E, X)) → bad-07(B))'

    def override_conjunction(info: OverrideConjunctionCallbackInfo) -> Clause | None:
        if info.instance_name != "b":
            return None
        antecedents = [*info.subterms]
        if info.closure_term:
            antecedents.append(info.closure_term)
        return Implies(And(*antecedents), info.predicate_term)

    def override_quantification(
        clause: Clause, info: OverrideQuantificationCallbackInfo
    ) -> Clause | None:
        if info.instance_name != "b":
            return None
        return Not(clause) if info.is_negated else clause

    override_scope_with_quantifiers_converter = AmrLogicConverter(
        existentially_quantify_instances=True,
        override_conjunction=override_conjunction,
        override_quantification=override_quantification,
    )
    with_quantifiers_logic = override_scope_with_quantifiers_converter.convert(amr_str)
    assert str(with_quantifiers_logic) == expected_with_quantifiers

    override_scope_without_quantifiers_converter = AmrLogicConverter(
        use_variables_for_instances=True,
        override_conjunction=override_conjunction,
        override_quantification=override_quantification,
    )
    without_quantifiers_logic = override_scope_without_quantifiers_converter.convert(
        amr_str
    )
    assert str(without_quantifiers_logic) == expected_without_quantifiers


def test_convert_amr_with_conditionals() -> None:
    amr_str = """
    (s / sing-01
        :ARG0 (b / boy)
        :condition (g / give-01
            :ARG1 (m / money)
            :ARG2 b))
    """
    expected = """
    ∃S(
        ∃B(
            ∃G(
                give-01(G) ∧
                ∃M(
                    :ARG1(G, M) ∧ money(M)
                ) ∧
                :ARG2(G, B) ∧
                boy(B)
            ) → (
                sing-01(S) ∧ :ARG0(S, B)
            )
        )
    )
    """
    implication_converter = AmrLogicConverter(
        existentially_quantify_instances=True,
        use_implies_for_conditions=True,
        use_variables_for_instances=True,
    )
    logic = implication_converter.convert(amr_str)
    assert fmt_logic(str(logic)) == fmt_logic(expected)
