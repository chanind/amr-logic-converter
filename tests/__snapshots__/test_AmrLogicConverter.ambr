# name: test_parse_amr_with_alignment_markers
  Exists(var=Var(name='e'), body=And(args=(Predicate(value='give-01', args=(Var(name='e'),), alignment=Alignment((2,))), Exists(var=Var(name='x'), body=And(args=(Predicate(value=':ARG0', args=(Var(name='e'), Var(name='x')), alignment=None), Predicate(value='person', args=(Var(name='x'),), alignment=None), Predicate(value=':named', args=(Var(name='x'), Const(value='"Ms Ribble"', alignment=Alignment((2,)))), alignment=None)))), Exists(var=Var(name='y'), body=And(args=(Predicate(value=':ARG2', args=(Var(name='e'), Var(name='y')), alignment=None), Predicate(value='child', args=(Var(name='y'),), alignment=Alignment((3,)))))), Exists(var=Var(name='z'), body=And(args=(Predicate(value=':ARG1', args=(Var(name='e'), Var(name='z')), alignment=None), Predicate(value='envelope', args=(Var(name='z'),), alignment=Alignment((4,)))))))))
# ---
